### PR TITLE
que3.3

### PR DESCRIPTION
**Pull request by ashok-choudhary-0**

### CODE REVIEW

File: models/user.js [hunk 1]
```diff
@@ -28,6 +28,10 @@ const user = sequelize.define('user', {
     type: DataTypes.STRING(100),
     allowNull: false
   },
+  jwtToken: {
```
Comment (amrit56789) [inline models/user.js]: Not jwtToken change a vairable name.

File: controllers/userController.js [hunk 2]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
-module.exports = { userRegister, login, getUserData, deleteUserData, limitUsersData, userAddress, deleteUserAddresses }
+
+const passwordResetToken = async (req, res) => {
+  const { user_id } = req.body
+  try {
+    const userData = await userModel.findOne({ where: { id: user_id } })
+    if (userData) {
+      const passwordResetToken = jwt.sign({
+        data: `${userData.user_id} ${userData.password}`
+      }, process.env.jwtSecKey, { expiresIn: 60 * 10 });
```
Comment (amrit56789) [inline controllers/userController.js:159]: Don't add expire time when you create token.

File: controllers/userController.js [hunk 3]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+      res.status(200).send(passwordResetToken)
+
+    } else {
+      res.status(404).send({ message: "user not found, please check the user id" })
+    }
+
+  } catch (err) { res.status(500).send(err) }
+
+}
+const verifyResetPassword = async (req, res) => {
```
Comment (amrit56789) [inline controllers/userController.js]: change a variable name you are checking password reset token

File: controllers/userController.js [hunk 4]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+  const resetToken = req.params.passwordResetToken
+  const { user_id, password } = req.body
+  try {
+    const userData = await userModel.findOne({ where: { id: user_id } })
+    const verifyToken = jwt.verify(resetToken, process.env.jwtSecKey)
```
Comment (amrit56789) [inline controllers/userController.js:177]: Here you are checking expiry time  and also token will be valid for 10 min

File: controllers/userController.js [hunk 5]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+    if (userData.jwtToken === resetToken && verifyToken) {
+      const hashedPassword = bcrypt.hashSync(password, 10);
+      await userModel.update({ password: hashedPassword, jwtToken: "" }, { where: { id: user_id } })
+      res.status(200).send({ message: "password updated successfully" })
+    } else {
+      res.status(200).send({ message: "password reset token expired" })
```
Comment (amrit56789) [inline controllers/userController.js]: Give a valid error.

File: controllers/userController.js [hunk 6]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+  } catch (err) { res.status(500).send(err) }
+
+}
+const verifyResetPassword = async (req, res) => {
+  const resetToken = req.params.passwordResetToken
+  const { user_id, password } = req.body
+  try {
+    const userData = await userModel.findOne({ where: { id: user_id } })
+    const verifyToken = jwt.verify(resetToken, process.env.jwtSecKey)
+    if (userData.jwtToken === resetToken && verifyToken) {
```
Comment (amrit56789) [inline controllers/userController.js]: and also here you can not directly reset the password if the token is valid than take a password if password is correct then create a hashpassword and update in the password field.

File: controllers/userController.js [hunk 7]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
     res.status(500).send({ message: err })
   }
 }
-module.exports = { userRegister, login, getUserData, deleteUserData, limitUsersData, userAddress, deleteUserAddresses }
+
+const passwordResetToken = async (req, res) => {
```
Comment (amrit56789) [inline controllers/userController.js]: This is a forgetPassword function not passwordResetToken

File: controllers/userController.js [hunk 8]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+    } else {
+      res.status(404).send({ message: "user not found, please check the user id" })
+    }
+
+  } catch (err) { res.status(500).send(err) }
+
+}
+const verifyResetPasswordToken = async (req, res) => {
+  const resetToken = req.params.passwordResetToken
+  const { username, new_password } = req.body
```
Comment (amrit56789) [inline controllers/userController.js]: Don't take username

File: controllers/userController.js [hunk 9]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+  try {
+    const userData = await userModel.findOne({ where: { username } })
```
Comment (amrit56789) [inline controllers/userController.js]: There is no need of userDate.

File: controllers/userController.js [hunk 10]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+    const verifyToken = jwt.verify(resetToken, process.env.jwtSecKey, { expiresIn: 60 * 60 })
```
Comment (amrit56789) [inline controllers/userController.js]: 60*60 is not a 10 min

File: controllers/userController.js [hunk 11]
```diff
@@ -147,4 +148,41 @@ const deleteUserAddresses = async (req, res) => {
...
+    if (userData.passwordResetToken === resetToken && verifyToken) {
```
Comment (amrit56789) [inline controllers/userController.js]: Here first you verify the token after that check password is coming or not, if password not present then take give a error, or if available then create a new password and update in the password field

File: index.js [hunk 12]
```diff
@@ -6,8 +6,6 @@ app.use(bodyParser.json())
 const rolesRouter = require("./router/rolesRouter");
 const userRouter = require("./router/userRouter");
```
Comment (amrit56789) [inline index.js:8]: Why you remove this?

File: controllers/userController.js [hunk 13]
```diff
@@ -147,4 +148,45 @@ const deleteUserAddresses = async (req, res) => {
...
+    const userData = await userModel.findOne({ where: { username } })
+    if (userData) {
+      const passwordResetToken = jwt.sign({
+        data: `${userData.username} ${userData.password}`
+      }, process.env.jwtSecKey);
+      await userModel.update({ passwordResetToken: passwordResetToken }, { where: { username } })
+      res.status(200).send(passwordResetToken)
+
+    } else {
+      res.status(404).send({ message: "user not found, please check the user id" })
```
Comment (amrit56789) [inline controllers/userController.js]: user_id?

File: controllers/userController.js [hunk 14]
```diff
@@ -147,4 +148,45 @@ const deleteUserAddresses = async (req, res) => {
...
+}
+const verifyResetPasswordToken = async (req, res) => {
+  const resetToken = req.params.passwordResetToken
+  const { new_password } = req.body
+  try {
+    const verifyToken = jwt.verify(resetToken, process.env.jwtSecKey, { expiresIn: 60 * 10 })
+    if(new_password){
+      if (verifyToken) {
+        const hashedPassword = bcrypt.hashSync(new_password, 10);
+        await userModel.update({ password: hashedPassword, passwordResetToken: "" }, { where: { passwordResetToken:resetToken } })
```
Comment (amrit56789) [inline controllers/userController.js]: Define in a variable and pass the update user data also take a variable name meaningfull.

File: controllers/userController.js [hunk 15]
```diff
@@ -147,4 +148,45 @@ const deleteUserAddresses = async (req, res) => {
...
 }
-module.exports = { userRegister, login, getUserData, deleteUserData, limitUsersData, userAddress, deleteUserAddresses }
+
+const forgotPassword = async (req, res) => {
+  const { username } = req.body
+  try {
+    const userData = await userModel.findOne({ where: { username } })
+    if (userData) {
+      const passwordResetToken = jwt.sign({
+        data: `${userData.username} ${userData.password}`
```
Comment (meanstack9716) [inline controllers/userController.js:158]: you didn't add the expiration time when creating the JWT token.  According to the assignment `Password reset token will only be active for 10 min.` @ashok-choudhary-0

File: controllers/userController.js [hunk 16]
```diff
@@ -147,4 +148,45 @@ const deleteUserAddresses = async (req, res) => {
...
+    }
+
+  } catch (err) { res.status(500).send(err) }
+
+}
+const verifyResetPasswordToken = async (req, res) => {
+  const resetToken = req.params.passwordResetToken
+  const { new_password } = req.body
+  try {
+    const verifyToken = jwt.verify(resetToken, process.env.jwtSecKey, { expiresIn: 60 * 10 })
```
Comment (meanstack9716) [inline controllers/userController.js]: are you sure? verify function accept expireIn? @ashok-choudhary-0 
according to jasonwebtoken  https://www.npmjs.com/package/jsonwebtoken there is no expireIn option

File: controllers/userController.js [hunk 17]
```diff
@@ -147,4 +148,45 @@ const deleteUserAddresses = async (req, res) => {
...
+    if (new_password) {
```
Comment (meanstack9716) [inline controllers/userController.js]: new_password this check should be on the top, before doing any thing. @ashok-choudhary-0

**Review by meanstack9716 — CHANGES_REQUESTED** [review]

**Pull request merged**